### ISSUE TITLE
"Task exception was never retrieved" with turn_off()

### ISSUE THREAD

**Issue by eibanez** [sbidy/pywizlight#42]
First of all, fantastic library! I'm only beginning to work with `asyncio` so forgive me if I'm missing something obvious here.

I am trying to use `asyncio.wait_for` to control the timeout of turning controlling some lights. I have created the following minimal example that tries to control a light bulb that is momentarily inaccessible (you can enter an invalid IP in the code below):

```
import asyncio

from pywizlight import wizlight, PilotBuilder

async def main():
    # Set up a standard light
    light = wizlight("192.168.0.170")
    
    # Set bulb brightness (with async timeout)
    timeout = 10
    
    try:
        #await asyncio.wait_for(light.turn_off(), timeout)
        await asyncio.wait_for(light.turn_on(PilotBuilder()), timeout)
    except asyncio.TimeoutError:
        print('Timeout ---------------')
    
    print('Sleep...')
    await asyncio.sleep(60)
    print('Done.')

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

When I use the `light.turn_on` method, there are no issues. The `try ... except` block works as expected and the output is as follows:

```
>python single.py
Timeout ---------------
Sleep...
Done.
```

However, when I use the `light.turn_off` method, asyncio complains:

```
>python single.py
Timeout ---------------
Sleep...
Task exception was never retrieved
future: <Task finished name='Task-4' coro=<wizlight.receiveUDPwithTimeout() done, defined at ...\Python38-32\lib\site
-packages\pywizlight\bulb.py:312> exception=TimeoutError()>
Traceback (most recent call last):
  File "...\Python38-32\lib\site-packages\pywizlight\bulb.py", line 314, in receiveUDPwithTimeout
    data, remote_addr = await asyncio.wait_for(stream.recv(), timeout)
  File "...\Python38-32\lib\asyncio\tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError
Done.
```

I see there are some internal `wait_for` calls in the library, but I have not been able to spot why `turn_on` works fine and `turn_off` does not.

Thanks for you help!

**Comment by sbidy** [comment]
Mh, I have to dig into that. Seems confusing 😉 ... give me some days.

**Comment by sbidy** [comment]
I'm not sure about the source of this issue but it seems to be a problem coming from the `receiveUDPwithTimeout` and `sendUDPMessage`. I don't now why this pops up "now" this is not an issue of the on/off function.
This will exception will be thrown it the bulb is offline.

**Comment by eibanez** [comment]
To be clear, I'm seeing this issue sporadically when the light bulb is on. The easiest way to reproduce the error is when the bulb is offline (or you use an incorrect IP).

I will take a look at the internals of the library to see why `turn_on` works fine but `turn_off` does not.

Are there any significant differences between the two?

**Comment by eibanez** [comment]
I guess an immediate difference is that `turn_on` uses `max_send_datagrams=10` but `turn_off` uses the default (`max_send_datagrams=100)`

```
    async def turn_off(self):
        """Turn the light off."""
        message = r'{"method":"setPilot","params":{"state":false}}'
        await self.sendUDPMessage(message)

    async def turn_on(self, pilot_builder=PilotBuilder()):
        """Turn the light on."""
        await self.sendUDPMessage(
            pilot_builder.get_pilot_message(), max_send_datagrams=10
        )
```

**Comment by eibanez** [comment]
Ok, I dug a little more into this. The difference between the `max_send_datagrams` does matter.

The `turn_off` command is stuck in the `while` loop, tries sending 19 times and ends there:
https://github.com/sbidy/pywizlight/blob/b9ef1e4eed17c3d322895898d804029cbccc3e57/pywizlight/bulb.py#L380-L388

The `turn_on` command exists that loop after trying 10 times and executes until this line:
https://github.com/sbidy/pywizlight/blob/b9ef1e4eed17c3d322895898d804029cbccc3e57/pywizlight/bulb.py#L390

I don't know enough about error catching in asyncio to know which `try... catch` is working and which is not. I activated the logger and the code never gets into this block. It seems to be captured by the `except` block I wrote.
https://github.com/sbidy/pywizlight/blob/b9ef1e4eed17c3d322895898d804029cbccc3e57/pywizlight/bulb.py#L393-L400

**Comment by sbidy** [comment]
Thank you for the research. It is little bit wired because there ware not major changes at all between older version (which are working) and the 0.4.1 / 0.4.2 I don't know why the exception are not catched by the try / catch block.

I'm also trying to find the source of this issue.

Is your bulb maybe placed in a spot where are the WiFi connections are bad or not reliable?

**Comment by eibanez** [comment]
I'm a new user, so I don't know if this happened before. Did you have a test that waited more than 30 seconds?

In the code above my `wait_for` times out at 10 s, before the internal `wait_for` in `receiveUDPwithTimeout` times out (30 seconds).

I think a solution would be ensuring that `max_send_datagrams * send_interval` does not exceed `timeout`. Otherwise, you are sending "datagrams" after you have stopped listening or waiting for them.

**Comment by sbidy** [comment]
Yes, this is the poin. I haven't time today to provide a fix.
This was also changed between the versions (from 60sec to 30sec).

Hopfully I can release a fixed version tomorrow.

**Comment by eibanez** [comment]
Thank you!

**Comment by sbidy** [comment]
It should be fixed with the 0.4.4 - the timeout is removed from the turn_on and turn_off.

**Comment by eibanez** [comment]
Thank you!
